exporters/otlp/otlphttp/driver.go

```diff
@@ -0,0 +1,269 @@
+		metricsURLPath: DefaultMetricsPath,
+		maxAttempts:    DefaultMaxAttempts,
+		backoff:        DefaultBackoff,
+	}
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+	for pathPtr, defaultPath := range map[*string]string{
+		&cfg.tracesURLPath:  DefaultTracesPath,
+		&cfg.metricsURLPath: DefaultMetricsPath,
+	} {
+		tmp := strings.TrimSpace(*pathPtr)
+		if tmp == "" {
+			tmp = defaultPath
+		} else {
+			tmp = path.Clean(tmp)
+			if !path.IsAbs(tmp) {
+				tmp = fmt.Sprintf("/%s", tmp)
+			}
+		}
+		*pathPtr = tmp
+	}
+	if cfg.maxAttempts <= 0 {
+		cfg.maxAttempts = DefaultMaxAttempts
+	}
+	if cfg.maxAttempts > DefaultMaxAttempts {
+		cfg.maxAttempts = DefaultMaxAttempts
+	}
+	if cfg.backoff <= 0 {
+		cfg.backoff = DefaultBackoff
+	}
+	client := &http.Client{}
+	if cfg.tlsCfg != nil {
+		transport := defaultTransportClone.Clone()
+		transport.TLSClientConfig = cfg.tlsCfg
+		client.Transport = transport
+	}
+	return &driver{
+		client: client,
+		cfg:    cfg,
+		stopCh: make(chan struct{}),
+	}
+}
+
+// Start implements otlp.ProtocolDriver.
+func (d *driver) Start(ctx context.Context) error {
+	// nothing to do
+	return nil
+}
+
+// Stop implements otlp.ProtocolDriver.
+func (d *driver) Stop(ctx context.Context) error {
+	close(d.stopCh)
+	return nil
+}
+
+// ExportMetrics implements otlp.ProtocolDriver.
+func (d *driver) ExportMetrics(ctx context.Context, cps metricsdk.CheckpointSet, selector metricsdk.ExportKindSelector) error {
+	rms, err := transform.CheckpointSet(ctx, selector, cps, 1)
+	if err != nil {
+		return err
+	}
+	if len(rms) == 0 {
+		return nil
+	}
+	pbRequest := &colmetricspb.ExportMetricsServiceRequest{
+		ResourceMetrics: rms,
+	}
+	rawRequest, err := pbRequest.Marshal()
+	if err != nil {
+		return err
+	}
+	return d.send(ctx, rawRequest, d.cfg.metricsURLPath)
+}
+
+// ExportTraces implements otlp.ProtocolDriver.
+func (d *driver) ExportTraces(ctx context.Context, ss []*tracesdk.SpanSnapshot) error {
+	protoSpans := transform.SpanData(ss)
+	if len(protoSpans) == 0 {
+		return nil
+	}
+	pbRequest := &coltracepb.ExportTraceServiceRequest{
+		ResourceSpans: protoSpans,
+	}
+	rawRequest, err := pbRequest.Marshal()
+	if err != nil {
+		return err
+	}
+	return d.send(ctx, rawRequest, d.cfg.tracesURLPath)
+}
+
+func (d *driver) send(ctx context.Context, rawRequest []byte, urlPath string) error {
+	address := fmt.Sprintf("%s://%s%s", d.getScheme(), d.cfg.address, urlPath)
+	var cancel context.CancelFunc
+	ctx, cancel = d.contextWithStop(ctx)
+	defer cancel()
+	for i := 0; i < d.cfg.maxAttempts; i++ {
+		response, err := d.singleSend(ctx, rawRequest, address)
+		if err != nil {
+			return err
+		}
+		// We don't care about the body, so close it
+		// immediately.
+		response.Body.Close()
+		switch response.StatusCode {
+		case http.StatusOK:
+			return nil
+		case http.StatusTooManyRequests:
+			fallthrough
+		case http.StatusServiceUnavailable:
+			select {
+			case <-time.After(getWaitDuration(d.cfg.backoff, i)):
```

Comment:
Should this respect the [`Retry-After`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After) header in addition to performing exponential backoff?

Reply:
Yeah, it should, but I decided to punt that to some follow-up PR.